File: src/main/java/parser/ast/expressions/ConditionalExpression.java
package parser.ast.expressions;

import stdlib.NumberValue;
import stdlib.StringValue;
import stdlib.Value;

public final class ConditionalExpression implements Expression {

    public enum Operator {

        EQUALS("=="),
        NOT_EQUALS("!="),

        LT("<"),
        LTEQ("<="),

        GT(">"),
        GTEQ(">="),

        AND("&&"),
        OR("||");

        private final String name;

        Operator(String name) {
            this.name = name;
        }

        public String getName() {
            if (name.equals("<")){
                return "&lt;";
            }
            if (name.equals(">")){
                return "&gt;";
            }
            if (name.equals("<=")){
                return "&lt;=";
            }
            if (name.equals(">=")){
                return "&gt;=";
            }
            if (name.equals("&&")){
                return "and";
            }
            return name;
        }

        @Override
        public String toString() {
            return String.format("<Operator operator=\"%s\"/>", getName());
        }
    }

    private Expression expr1, expr2;
    private Operator operation;

    public ConditionalExpression(Operator operation, Expression expr1, Expression expr2) {
        this.operation = operation;
        this.expr1 = expr1;
        this.expr2 = expr2;
    }

    @Override
    public Value eval() {
        final var value1 = expr1.eval();
        final var value2 = expr2.eval();

        double number1, number2;
        if (value1 instanceof StringValue) {
            number1 = value1.asString().compareTo(value2.asString());
            number2 = 0;
        } else {
            number1 = value1.asNumber();
            number2 = value2.asNumber();
        }


        boolean result;
        switch (operation) {
            case LT: result = number1 < number2; break;
            case LTEQ: result = number1 <= number2; break;
            case GT: result = number1 > number2; break;
            case GTEQ: result = number1 >= number2; break;
            case NOT_EQUALS: result = number1 != number2; break;
            case AND: result = (number1 != 0) && (number2 != 0); break;
            case OR: result = (number1 != 0) || (number2 != 0); break;
            case EQUALS:
            default:
                result = number1 == number2; break;
        }
        return new NumberValue(result);
    }

    @Override
    public String toString() {
        return String.format("<ConditionalExpression>%s%s%s</ConditionalExpression>", expr1, operation.toString(), expr2);

    }
}

File: src/main/java/parser/ast/expressions/ConditionalExpression.kt
package parser.ast.expressions

import stdlib.NumberValue
import stdlib.StringValue
import stdlib.Value

class ConditionalExpression(private val operation: Operator, private val expr1: Expression, private val expr2: Expression) : Expression {

    enum class Operator private constructor(private val namee: String) {

        EQUALS("=="),
        NOT_EQUALS("!="),

        LT("<"),
        LTEQ("<="),

        GT(">"),
        GTEQ(">="),

        AND("&&"),
        OR("||");

        fun getNamee(): String {
            return when (namee) {
                "<" -> "&lt;"
                ">" -> "&gt;"
                "<=" -> "&lt;="
                ">=" -> "&gt;="
                "&&" -> "and"
                else -> namee
            }
        }

        override fun toString(): String {
            return String.format("<Operator operator=\"%s\"/>", getNamee())
        }
    }

    override fun eval(): Value {
        val value1 = expr1.eval()
        val value2 = expr2.eval()

        val number1: Double
        val number2: Double
        if (value1 is StringValue) {
            number1 = value1.asString().compareTo(value2.asString()).toDouble()
            number2 = 0.0
        } else {
            number1 = value1.asNumber()
            number2 = value2.asNumber()
        }


        val result: Boolean
        result = when (operation) {
            ConditionalExpression.Operator.LT -> number1 < number2
            ConditionalExpression.Operator.LTEQ -> number1 <= number2
            ConditionalExpression.Operator.GT -> number1 > number2
            ConditionalExpression.Operator.GTEQ -> number1 >= number2
            ConditionalExpression.Operator.NOT_EQUALS -> number1 != number2
            ConditionalExpression.Operator.AND -> number1 != 0.0 && number2 != 0.0
            ConditionalExpression.Operator.OR -> number1 != 0.0 || number2 != 0.0
            ConditionalExpression.Operator.EQUALS -> number1 == number2
        }
        return NumberValue(result)
    }

    override fun toString(): String {
        return String.format("<ConditionalExpression>%s%s%s</ConditionalExpression>", expr1, operation.toString(), expr2)

    }
}
